build/deploy/main.go

```diff
@@ -34,11 +34,29 @@ func deploy() error {
 	bundlePath := os.Args[2]
 
 	siteURL := os.Getenv("MM_SERVICESETTINGS_SITEURL")
+	adminToken := os.Getenv("MM_ADMIN_TOKEN")
 	adminUsername := os.Getenv("MM_ADMIN_USERNAME")
 	adminPassword := os.Getenv("MM_ADMIN_PASSWORD")
 	copyTargetDirectory, _ := filepath.Abs("../mattermost-server")
-	if siteURL != "" && adminUsername != "" && adminPassword != "" {
-		return uploadPlugin(pluginID, bundlePath, siteURL, adminUsername, adminPassword)
+
+	if siteURL != "" {
+		client := model.NewAPIv4Client(siteURL)
+
+		if adminToken != "" {
+			log.Printf("Authenticating using token against %s.", siteURL)
+			client.SetToken(adminToken)
+
+			return uploadPlugin(client, pluginID, bundlePath)
+		} else if adminUsername != "" && adminPassword != "" {
+			client := model.NewAPIv4Client(siteURL)
+			log.Printf("Authenticating as %s against %s.", adminUsername, siteURL)
+			_, resp := client.Login(adminUsername, adminPassword)
+			if resp.Error != nil {
+				return fmt.Errorf("Failed to login as %s: %s", adminUsername, resp.Error.Error())
```

Comment:
We should wrap the error
```suggestion
				return errors.Wrap(err, "Failed to login as %", adminUsername)
```